### PR TITLE
[core] Introduce a flag which allows a longer timeout for raylet when GCS restarts.

### PR DESCRIPTION
**Pull request by fishbone**
Signed-off-by: Yi Cheng <chengyidna@gmail.com>

## Why are these changes needed?
When GCS restarts, sometimes, raylet needs a while to reconnect to the GCS, for example, in k8s env, it needs a while to move GSC to the service. This PR try to fix this by allowing a longer timeout for the first ping when GCS restarts.

Once GCS get the first ping, it'll just use the regular timeout instead.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by brucez-anyscale** [comment]
lgtm

File: src/ray/gcs/gcs_server/gcs_heartbeat_manager.h [hunk 1]
```diff
@@ -89,7 +94,10 @@ class GcsHeartbeatManager : public rpc::HeartbeatInfoHandler {
   /// The callback of node death.
   std::function<void(const NodeID &)> on_node_death_callback_;
   /// The number of heartbeats that can be missed before a node is removed.
-  int64_t num_heartbeats_timeout_;
+  const int64_t num_heartbeats_timeout_;
+  /// The initial heartbeat timeout when GCS restarts and waiting for raylet to
+  /// reconnect. Once connected, we'll use num_heartbeats_timeout_
+  const int64_t initial_num_heartbeats_timeout_;
```
Comment (scv119) [inline src/ray/gcs/gcs_server/gcs_heartbeat_manager.h:100]: the name is a bit confusing, how about?  `gcs_failover_worker_reconnect_timeout_` ?

File: src/ray/common/ray_config_def.h [hunk 2]
```diff
@@ -57,6 +57,11 @@ RAY_CONFIG(uint64_t, gcs_pull_resource_loads_period_milliseconds, 1000)
 /// heartbeat intervals, the raylet monitor process will report
 /// it as dead to the db_client table.
 RAY_CONFIG(int64_t, num_heartbeats_timeout, 30)
+
+/// If GCS restarts, before the first heatbeat is sent, initial_num_heartbeats_timeout
```
Comment (scv119) [inline src/ray/common/ray_config_def.h:61]: we can add a bit detail why this has to be different from the default timeout value.

**Review by scv119 — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_heartbeat_manager.cc [hunk 3]
```diff
@@ -41,7 +45,7 @@ GcsHeartbeatManager::GcsHeartbeatManager(
 void GcsHeartbeatManager::Initialize(const GcsInitData &gcs_init_data) {
   for (const auto &item : gcs_init_data.Nodes()) {
     if (item.second.state() == rpc::GcsNodeInfo::ALIVE) {
-      AddNode(item.second);
+      AddNode(item.second, initial_num_heartbeats_timeout_);
```
Comment (scv119) [inline src/ray/gcs/gcs_server/gcs_heartbeat_manager.cc:48]: would a private `AddReconnectingNode` function make it clearer?

**Pull request merged**